<a href="https://colab.research.google.com/github/Carolinsrainbow/UC_AnalisisdeDatosEnPython/blob/main/Laboratorio2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Laboratorio 2

Bienvenido al segundo laboratorio del curso Análisis de Datos en Python.

Escribe el código correspondiente para contestar cada una de las preguntas. Fíjate muy bien en el nombre de las variables, ya que de lo contrario, tu trabajo no podrá ser evaluado de manera correcta.

Debes ir ejecutando las celdas de codigo en el orden que aparecen, para evitar tener errores. No elimines celdas!!

Este trabajo tiene un límite de tiempo que puedes visualizar al inicio de esta pantalla.

### Pregunta 1
¿Cuál es el promedio y desviación estándar de la cantidad de basura diaria generada en cada calle?

Indicaciones:
- Guarda los datos en un dataframe, bajo la variable "basuraxcalle"
- El dataframe debe tener 3 columnas: "Calle", "Cantidad promedio [kilos de basura]" y "Desv est [kilos de basura]"

In [4]:
import pandas as pd
import numpy as np

ciudad = pd.read_excel('Datos.xlsx',sheet_name = 'Ciudad')
basura = pd.read_excel('Datos.xlsx',sheet_name = 'Histórico de basura')

basuraxcalle= None

# Calcular el promedio y la desviación estándar de la cantidad de basura diaria por calle
basuraxcalle = basura.groupby('Calle')['Cantidad [kilos de basura]'].agg(['mean', 'std']).reset_index()

# Renombrar las columnas según los requisitos
basuraxcalle.columns = ['Calle', 'Cantidad promedio [kilos de basura]', 'Desv est [kilos de basura]']

# Mostrar el resultado
print(basuraxcalle)

           Calle  Cantidad promedio [kilos de basura]  \
0        Calle 1                             2.982143   
1       Calle 10                             3.192857   
2      Calle 100                             2.664286   
3     Calle 1000                             2.742857   
4     Calle 1001                             2.896429   
...          ...                                  ...   
3235   Calle 995                             3.032143   
3236   Calle 996                             2.203571   
3237   Calle 997                             2.964286   
3238   Calle 998                             2.978571   
3239   Calle 999                             2.867857   

      Desv est [kilos de basura]  
0                       1.209240  
1                       1.279447  
2                       1.199449  
3                       1.211191  
4                       1.093578  
...                          ...  
3235                    1.172858  
3236                    0.850918  


Revisemos que el dataframe haya quedado bien creado, la siguiente celda no tiene puntaje, pero te servirá para que puedas revisar que hayas creado bien tu dataframe

In [15]:
assert list(basuraxcalle.columns) == ['Calle', 'Cantidad promedio [kilos de basura]', 'Desv est [kilos de basura]'], "Los nombres de las columnas son incorrectos"

In [16]:
basuraxcalle[basuraxcalle['Calle'] == "Calle 1"]

Calle  Cantidad promedio [kilos de basura]  Desv est [kilos de basura]
0  Calle 1                             2.982143                     1.20924

El resultado para la Calle 1 deberia ser


|Calle   | Cantidad promedio [kilos de basura]	  | Desv est [kilos de basura]  |
|---|---|---|
|Calle 1  | 2.982143  | 1.20924  |

No es necesario que redondees el numero de promedio o desviacion estandar, ni que hagas ninguna operación con el: guardalo en el dataframe tal cual lo entregue el calculo.

Revisemos que hayas calculado bien el promedio y desviacion para la calle 1:

In [17]:
assert np.round(basuraxcalle[basuraxcalle['Calle'] == "Calle 1"]["Cantidad promedio [kilos de basura]"].values,2) == [[2.98]], "Fallo en el promedio"
assert np.round(basuraxcalle[basuraxcalle['Calle'] == "Calle 1"]["Desv est [kilos de basura]"].values,2) == [[1.21]], "Fallo en la desviacion estandar"

La siguiente celda hace tests ocultos, en el mismo formato del anterior pero para otras calles seleccionadas al azar! No modifiques, muevas ni elimines la celda

### Pregunta 2

Se define como **calle “intensa”** a aquellas calles cuyo promedio diario de basura generada es **estrictamente mayor a 2,92 kilos por día**.

¿Cuántas de las calles serían consideradas intensas (en función de los datos históricos de 28 días)?

Guarda la cantidad de calles intensas en la variable c_intensas, en formato int.

NO DEBES REDONDEAR NI APROXIMAR EL PROMEDIO PARA HACER LA CLASIFICACION!!

**Hint: usa el dataframe que construiste para la pregunta anterior!**

In [18]:
c_intensas=0

# your code here
#raise NotImplementedError

#En base a la anterior, se saca el filtro de mayor estricto que 2.92
calles_intensas = basuraxcalle[basuraxcalle['Cantidad promedio [kilos de basura]'] >= 2.92]

# Contar la cantidad de calles intensas
c_intensas = int(calles_intensas.shape[0])
c_intensas

1657

La siguiente celda hace tests ocultos. No modifiques, muevas ni elimines la celda

### Pregunta 3

En el dataframe `ciudad` hay datos de las coordenadas donde comienzan y terminan las calles.

La comuna comienza en la coordenada 0 y termina en la coordenada 40, tanto para X como para Y. Si se divide territorialmente la comuna de la siguiente manera:

#### a. En X, desde la coordenada 0 a la 22 (inclusive) se define Oeste (O)
#### b. En X, desde la coordenada 23 a la 40 (inclusive) se define Este (E)
#### c. En Y, desde la coordenada 0 a la 27 (inclusive) se define Norte (N)
#### d. En Y, desde la coordenada 28 a la 40 (inclusive) se define Sur (S)

Considerando esta división territorial, las calles de la comuna se clasifican como zonas: N-E, N-O, S-E, S-O.

Puede suceder que alguna calle comience en una zona y termine en otra. Veamos un ejemplo!

Primero renombremos las columnas

In [19]:
ciudad = ciudad.rename(columns = {'Se considera que la ciudad comienza en la esquina inferior izquierda de la cuadrícula que la representa, y entonces: X=0 e Y=0':'Calle',
          'Coordenada (X,Y) donde comienza la calle':'Coordenada inicial',
       'Coordenada (X,Y) donde termina la calle':'Coordenada final'})

Ahora veamos por ejemplo la calle 28

In [20]:
ciudad[ciudad["Calle"] == "Calle 28"]

Calle Coordenada inicial Coordenada final  Unnamed: 3 Unnamed: 4
27  Calle 28             (0,27)           (0,28)         NaN        NaN

Esta calle comienza en la coordenada (0,27), pero termina en (0,28). Esto quiere decir, que empieza en la zona N-O, pero termina en la zona S-O. Esta es una clase que no se puede clasificar!!

A continuacion, guarda en la variable `sin_clasificar` la cantidad de calles que no pueden ser clasificadas en una zona, es decir, que comiencen en una y terminen en otra. El formato debe ser int

In [29]:
sin_clasificar=None
# your code here
#raise NotImplementedError

def determinar_zona(coordenada):
    x, y = coordenada
    if x <= 22:
        x_zona = 'O'
    else:
        x_zona = 'E'

    if y <= 27:
        y_zona = 'N'
    else:
        y_zona = 'S'

    return y_zona + '-' + x_zona

# Verificar y convertir coordenadas iniciales y finales a tuplas de enteros
ciudad['Coordenada inicial'] = ciudad['Coordenada inicial'].apply(lambda x: tuple(map(int, x.strip('()').split(','))) if isinstance(x, str) else x)
ciudad['Coordenada final'] = ciudad['Coordenada final'].apply(lambda x: tuple(map(int, x.strip('()').split(','))) if isinstance(x, str) else x)

# Determinar las zonas para coordenadas iniciales y finales
ciudad['Zona inicial'] = ciudad['Coordenada inicial'].apply(determinar_zona)
ciudad['Zona final'] = ciudad['Coordenada final'].apply(determinar_zona)

# Contar las calles que no pueden ser clasificadas en una sola zona
sin_clasificar = ciudad[ciudad['Zona inicial'] != ciudad['Zona final']].shape[0]

# Mostrar la cantidad de calles sin clasificar
print(sin_clasificar)

82


La siguiente celda hace tests ocultos. No modifiques, muevas ni elimines la celda

### Pregunta 4
Considerando la clasificación anterior,
¿cuántas calles quedan clasificadas en cada zona?

Para contestar, genera un dataframe que tenga 2 columnas: Clasificacion y Calle. Dicho df debe ser almacenado en la variable "callesxzona".

Los valores para clasificacion son: "N-E", "N-O", "No clasificado", "S-E", "S-O"

Ejemplo del dataframe:

|Clasificacion|Calle|
|---|---|
|N-E| 0|
|N-O| 0|
|No clasificado| 0|
|S-E| 0|
|S-O| 0|



In [30]:
callesxzona=None
# your code here
#raise NotImplementedError




    Clasificacion  Calle
0             N-E    962
1             N-O   1210
2             S-E    437
3             S-O    549
4  No clasificado     82


In [36]:
# Crear el DataFrame para la clasificación de calles por zona
ciudad['Clasificacion'] = ciudad.apply(lambda row: row['Zona inicial'] if row['Zona inicial'] == row['Zona final'] else 'No clasificado', axis=1)

# Contar cuántas calles hay en cada clasificación
callesxzona = ciudad['Clasificacion'].value_counts().reset_index()
callesxzona.columns = ['Clasificacion', 'Calle']

# Asegurar que todas las clasificaciones están presentes en el DataFrame
clasificaciones = ['N-E', 'N-O', 'S-E', 'S-O', 'No clasificado']
for clas in clasificaciones:
    if clas not in callesxzona['Clasificacion'].values:
        callesxzona = callesxzona.append({'Clasificacion': clas, 'Calle': 0}, ignore_index=True)

# Ordenar el DataFrame por clasificación
callesxzona = callesxzona.set_index('Clasificacion').reindex(clasificaciones).reset_index()


Revisa que hayas creado bien el dataframe!

In [37]:
print(list(callesxzona.columns) == ['Clasificacion', 'Calle'])

clasificaciones = list(callesxzona["Clasificacion"].values)
clasificaciones.sort()

print(clasificaciones == ['N-E', 'N-O', 'No clasificado', 'S-E', 'S-O'])

True
True


Comenzamos con un test visible!

In [38]:
assert callesxzona[callesxzona['Clasificacion'] == 'N-E']["Calle"].values == [[962]]

La siguiente celda hace tests ocultos. No modifiques, muevas ni elimines la celda

### Pregunta 5
¿Cuál es el promedio y desviación estándar de la cantidad de kilos de basura generada por cada zona?

Excluya las calles que quedan sin clasificar.

Guarde su respuesta en un dataframe con la siguiente estructura:

|Zona   | Cantidad promedio [kilos de basura]	  | Desv est [kilos de basura]  |
|---|---|---|
|N-O  | 0  | 0  |

Guarde el dataframe en la variable `zona_stats`

In [41]:
zona_stats=None
# your code here
# raise NotImplementedError

# Crear el DataFrame para la clasificación de calles por zona
ciudad['Clasificacion'] = ciudad.apply(lambda row: row['Zona inicial'] if row['Zona inicial'] == row['Zona final'] else 'No clasificado', axis=1)

# Excluir calles no clasificadas
ciudad_clasificada = ciudad[ciudad['Clasificacion'] != 'No clasificado']

# Combinar con los datos de basura
basura = basura.merge(ciudad_clasificada[['Calle', 'Clasificacion']], on='Calle', how='inner')

# Calcular el promedio y la desviación estándar por zona
zona_stats = basura.groupby('Clasificacion')['Cantidad [kilos de basura]'].agg(['mean', 'std']).reset_index()

# Renombrar columnas
zona_stats.columns = ['Zona', 'Cantidad promedio [kilos de basura]', 'Desv est [kilos de basura]']

# Asegurar que todas las zonas están presentes en el DataFrame
zonas = ['N-E', 'N-O', 'S-E', 'S-O']
for zona in zonas:
    if zona not in zona_stats['Zona'].values:
        zona_stats = zona_stats.append({'Zona': zona, 'Cantidad promedio [kilos de basura]': 0, 'Desv est [kilos de basura]': 0}, ignore_index=True)

# Ordenar el DataFrame por zona
zona_stats = zona_stats.set_index('Zona').reindex(zonas).reset_index()

# Mostrar el resultado
print(zona_stats)

  Zona  Cantidad promedio [kilos de basura]  Desv est [kilos de basura]
0  N-E                             2.926032                    1.122461
1  N-O                             2.920708                    1.123545
2  S-E                             2.931636                    1.116296
3  S-O                             2.928220                    1.117884


Revisemos el formato

In [42]:
zona_stats


Zona  Cantidad promedio [kilos de basura]  Desv est [kilos de basura]
0  N-E                             2.926032                    1.122461
1  N-O                             2.920708                    1.123545
2  S-E                             2.931636                    1.116296
3  S-O                             2.928220                    1.117884

In [43]:
print(list(zona_stats.columns) == ['Zona', 'Cantidad promedio [kilos de basura]', 'Desv est [kilos de basura]'])

cols = list(zona_stats['Zona'].values)
cols.sort()

print(cols == ['N-E', 'N-O', 'S-E', 'S-O'])

True
True


Comenzamos con un test visible:

In [44]:
ans_NE = np.round(zona_stats[zona_stats["Zona"]=='N-E'][["Cantidad promedio [kilos de basura]", "Desv est [kilos de basura]"]].values,3)

assert (ans_NE== [[2.926, 1.122]]).all()

La siguiente celda hace tests ocultos. No modifiques, muevas ni elimines la celda